File: actin-common/src/main/java/com/hartwig/actin/clinical/feed/LabEntry.java
package com.hartwig.actin.clinical.feed;

import java.time.LocalDate;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class LabEntry {

    @NotNull
    public abstract String subject();

    @NotNull
    public abstract String codeCodeOriginal();

    @NotNull
    public abstract String codeDisplayOriginal();

    @NotNull
    public abstract LocalDate issued();

    @NotNull
    public abstract String valueQuantityComparator();

    public abstract double valueQuantityValue();

    @NotNull
    public abstract String valueQuantityUnit();

    @NotNull
    public abstract String interpretationDisplayOriginal();

    @Nullable
    public abstract String valueString();

    @Nullable
    public abstract String codeCode();

    @NotNull
    public abstract String referenceRangeText();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/lab/LabEntry.kt
package com.hartwig.actin.clinical.feed.emc.lab

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hartwig.actin.datamodel.clinical.provided.JacksonSerializable
import com.hartwig.actin.clinical.feed.emc.EuropeanDecimalDeserializer
import com.hartwig.actin.clinical.feed.emc.FeedEntry
import com.hartwig.actin.clinical.feed.emc.FeedSubjectDeserializer
import java.time.LocalDate

@JacksonSerializable
data class LabEntry(
    @JsonProperty("subject")
    @JsonDeserialize(using = FeedSubjectDeserializer::class)
    override val subject: String,

    @JsonProperty("code_code_original")
    val codeCodeOriginal: String,

    @JsonProperty("code_display_original")
    val codeDisplayOriginal: String,

    @JsonProperty("valueQuantity_comparator")
    val valueQuantityComparator: String,

    @JsonProperty("valueQuantity_value")
    @JsonDeserialize(using = EuropeanDecimalDeserializer::class)
    val valueQuantityValue: Double,

    @JsonProperty("valueQuantity_unit")
    val valueQuantityUnit: String,

    @JsonProperty("referenceRange_text")
    val referenceRangeText: String,

    @JsonProperty("effectiveDateTime")
    val effectiveDateTime: LocalDate
) : FeedEntry